shared/src/ledger/native_vp/masp.rs

```diff
@@ -161,6 +166,43 @@ where
                     return Ok(false);
                 }
             }
+
+            let mut revealed_nullifiers = HashSet::new();
+            for description in shielded_tx
+                .sapling_bundle()
+                .map_or(&vec![], |description| &description.shielded_spends)
+            {
+                let nullifier_key = Key::from(MASP.to_db_key())
+                    .push(&MASP_NULLIFIERS_KEY_PREFIX.to_owned())
+                    .expect("Cannot obtain a storage key")
+                    .push(&namada_core::types::hash::Hash(
+                        description.nullifier.0,
+                    ))
+                    .expect("Cannot obtain a storage key");
+                if self.ctx.has_key_pre(&nullifier_key)?
+                    || revealed_nullifiers.contains(&nullifier_key)
+                {
+                    tracing::debug!(
+                        "MASP double spending attempt, the nullifier {:#?} \
+                         has already been revealed previously",
+                        description.nullifier.0
+                    );
+                    return Ok(false);
+                }
+                revealed_nullifiers.insert(nullifier_key);
```

Comment:
Right now in the VP we check that the nullifier keys changed are a subset of those in the transaction. But where do we check that there is a key change for each of the nullifiers in the transaction? I would expect this check to be somewhere in this loop...

Reply:
Nice catch thanks! I've added the missing check now